accepted/future-releases/records/records-feature-specification.md

```diff
@@ -586,8 +586,51 @@ The runtime type of `pair` is `(int, double)`, not `(num, Object)`, However, the
 variable declaration is still valid and sound because records are naturally
 covariant in their field types.
 
+### Interactions with libraries using older language versions
+
+The records feature is language versioned, as usual with new Dart features.
+This means that it will be an error to use the syntax for records in libraries
+which do not have a language version greater than or equal to the language
+version in which records are released.  More specifically, assuming that `v` is
+the language version in which records are released, the following errors apply.
+
+It is an error to reference the SDK class `Record` in a library whose language
+version is less than `v`.
```

Comment:
> Just drop this paragraph entirely, it's not a restriction we need.

I don't think this is true.  I can clarify this, I see no issues with specifying this more precisely if you feel that the intent is unclear, but this is an important error.  Without this error, you can write a library which uses the `Record` class (which will work with no errors or warnings, since you are running on an sdk that supports it) and then ship the code to pub without raising your sdk constraint, whereupon you break the world.

> The Record declaration will be marked with @Since("v"), so we can still give warnings if you use it in a package with min-SDK < v.
(And if your min SDK >= v anyway, there is no problem, even if the referencing library is held back by a //@dart=<v marker.)

Empirically, the `@Since` annotation doesn't seem to work this way.  I can write a package with sdk constraint   `sdk: ">=2.9.0 <3.0.0"` and use the `Object.hash` method, which is marked `@Since("2.14")` with  no errors or warnings when running with a bleeding edge SDK.  

> To disallow direct references to the Record declaration, we'd have to be explicit and, say, state that you are not allowed to have the identifier Record in a place where it resolves to/denotes the Record class declaration of dart:core.

This is more restrictive than we need (it's fine to use the identifier `Record`, which resolves to the `Record` class from `dart:core`, if it is re-exported via another library).  But I don't know that we care about this use case.  I'll update this with more precise wording, either using this restriction or something more fine grained.

Reply:
I expecting to not restrict access to the type. We don't generally restrict access to new features in the platform libraries to libraries with language versions before the SDK where the feature was introduced.

We could, and then it should really apply to every `@Since`-annotated platform declaration, but so far we haven't done that. So, is there something which makes `Record` special in this regard, or is it something we would want to apply more generally?


As for phrasing, if we do want the restriction, maybe:
> ... must not contain the identifier `Record` in a place where the name denotes the `Record` class declaration of `dart:core`, and where that import scope name is only imported from platform libraries.

That should allow any reference which is also exported through a non-platform library.

Reply:
> ... must not contain the identifier Record in a place where the name denotes the Record class declaration of dart:core, and where that import scope name is only imported from platform libraries.

I used something modeled after this.